src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -0,0 +1,256 @@
+  }
+  
+  /** If `p` is a parameter of `pt`, propagate the non-parameter bounds
+   *  of `p` to the parameters known to be less or greater than `p`.
+   */
+  def initialize(pt: PolyType): Boolean =
+    checkPropagated(i"initialized $pt") {
+      pt.paramNames.indices.forall { i =>
+        val param = PolyParam(pt, i)
+        val bounds = constraint.nonParamBounds(param)
+        val lower = constraint.lower(param)
+        val upper = constraint.upper(param)
+        if (lower.nonEmpty && !bounds.lo.isRef(defn.NothingClass) ||
+          upper.nonEmpty && !bounds.hi.isRef(defn.AnyClass)) println(i"INIT*** $pt")
+        lower.forall(addOneBound(_, bounds.hi, isUpper = true)) &&
+          upper.forall(addOneBound(_, bounds.lo, isUpper = false))
+      }
+    }
+
+  /** Can `param` be constrained with new bounds? */
+  final def canConstrain(param: PolyParam): Boolean =
+    !frozenConstraint && (constraint contains param)
+
+  /** Add constraint `param <: bond` if `fromBelow` is true, `param >: bound` otherwise.
+   *  `bound` is assumed to be in normalized form, as specified in `firstTry` and
+   *  `secondTry` of `TypeComparer`. In particular, it should not be an alias type,
+   *  lazy ref, typevar, wildcard type, error type. In addition, upper bounds may
+   *  not be AndTypes and lower bounds may not be OrTypes.
+   */
+  protected def addConstraint(param: PolyParam, bound: Type, fromBelow: Boolean): Boolean = {
+    def description = i"constr $param ${if (fromBelow) ">:" else "<:"} $bound:\n$constraint"
+    //checkPropagated(s"adding $description")(true) // DEBUG in case following fails
+    checkPropagated(s"added $description") {
+      addConstraintInvocations += 1
+      try bound match {
+        case bound: PolyParam if constraint contains bound =>
+          if (fromBelow) addLess(bound, param) else addLess(param, bound)
+        case _ =>
+          if (fromBelow) addLowerBound(param, bound) else addUpperBound(param, bound)
+      }
+      finally addConstraintInvocations -= 1
+    }
+  }
+   
+  /** Check that constraint is fully propagated. See comment in Config.checkConstraintsPropagated */
+  def checkPropagated(msg: => String)(result: Boolean): Boolean = {
+    if (Config.checkConstraintsPropagated && result && addConstraintInvocations == 0) {
+      frozenConstraint = true
+      for (p <- constraint.domainParams) {
+        def check(cond: => Boolean, q: PolyParam, ordering: String, explanation: String): Unit =
+          assert(cond, i"propagation failure for $p $ordering $q: $explanation\n$msg")
+        for (u <- constraint.upper(p))
+          check(bounds(p).hi <:< bounds(u).hi, u, "<:", "upper bound not propagated")
+        for (l <- constraint.lower(p)) {
+          check(bounds(l).lo <:< bounds(p).hi, l, ">:", "lower bound not propagated")
+          check(constraint.isLess(l, p), l, ">:", "reverse ordering (<:) missing")
+        }
+      }
+      frozenConstraint = false
```

Comment:
I guess this should not set to false, but instead set to previously stored value